mobilecoind-dev-faucet/src/data_types.rs

```diff
@@ -188,3 +196,149 @@ impl From<Result<api::SubmitTxResponse, String>> for JsonSubmitTxResponse {
         }
     }
 }
+
+/// A json request from a user to initiate a slam
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct JsonSlamRequest {
+    /// Target num txs to submit in the slam
+    pub target_num_tx: Option<u32>,
+    /// Number of threads to create during slamming
+    pub num_threads: Option<u32>,
+    /// Number of retries to use when submitting Txs
+    pub retries: Option<u32>,
+    /// How much ahead of the network to set the tombstone block
+    pub tombstone_offset: Option<u32>,
+    /// Which consensus endpoints to submit transactions to
+    pub consensus_uris: Option<Vec<String>>,
+}
+
+// Construct SlamParams from JsonSlamRequest, using defaults to fill in any
+// omitted values.
+impl TryFrom<&JsonSlamRequest> for SlamParams {
+    type Error = String;
+    fn try_from(req: &JsonSlamRequest) -> Result<SlamParams, String> {
+        let mut result = SlamParams::default();
+        if let Some(val) = req.target_num_tx.as_ref() {
+            result.target_num_tx = *val;
+        }
+        if let Some(val) = req.num_threads.as_ref() {
+            result.num_threads = *val;
+        }
+        if let Some(val) = req.retries.as_ref() {
+            result.retries = *val;
+        }
+        if let Some(val) = req.tombstone_offset.as_ref() {
+            result.tombstone_offset = *val;
+        }
+        if let Some(uris) = req.consensus_uris.as_ref() {
+            result.consensus_client_uris = uris
+                .iter()
+                .map(|uri| FromStr::from_str(uri.as_str()))
+                .collect::<Result<Vec<_>, _>>()
+                .map_err(|err| format!("Invalid uri: {}", err))?;
+        }
+        Ok(result)
+    }
+}
+
+/// A slam resposne includes the parameters used to start the slam, and the
+/// report at the end
+#[derive(Debug, Default)]
+pub struct SlamResponse {
+    /// The slam params actually used
+    pub params: SlamParams,
+    /// The report for the slam operation
+    pub report: SlamReport,
+}
+
+/// Json form of slam params
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub struct JsonSlamParams {
+    /// The target number of txs to submit
+    pub target_num_tx: u32,
+    /// The number of threads to use to submit txs in parallel
+    pub num_threads: u32,
+    /// The number of retries when submitting a transaction
+    pub retries: u32,
+    /// How long to wait before retrying (seconds)
+    pub retry_period: f32,
```

Comment:
It's a bit of a nit but pretty much everywhere else in the codebase we've used milliseconds instead of seconds-as-floats.

Reply:
Yea that's probly better